File: src/application/strategies/factories/strategy_application_factory.py
from collections.abc import Callable, Mapping
from enum import Enum
from application.strategies.langchain_strategy import LangChainStrategy
from domain.abstracts.domain_type import DomainType
from application.strategies.abstracts.base_application_strategy import BaseApplicationStrategy
from application.strategies.claude_strategy import ClaudeStrategy
from application.strategies.gemini_strategy import GeminiStrategy
from application.strategies.groq_strategy import GroqStrategy
from application.strategies.openai_strategy import OpenAIStrategy

class StrategyApplicationFactory:
    
    @property
    def default(self) -> DomainType:
        return self._default
    def __init__(self,
        creators: Mapping[DomainType, Callable[[], BaseApplicationStrategy]] | None = None,
        default: DomainType = DomainType.Groq) -> None:

        self._creators: dict[DomainType, Callable[[], BaseApplicationStrategy]] = dict(creators) if creators else {
            DomainType.Claude: ClaudeStrategy,
            DomainType.OpenAI: OpenAIStrategy,
            DomainType.Gemini: GeminiStrategy,
            DomainType.Groq: GroqStrategy,
            DomainType.LangChain: LangChainStrategy
        }

        self._default = default

        if self._default not in self._creators:
            raise ValueError("A strategy padrão precisa existir no catálogo")

    def _parse_domain(self, name: str | DomainType | None) -> DomainType:
        # Já é DomainType → retorna direto (fast path)
        if isinstance(name, DomainType):
            return name

        # None ou vazio → default
        if not name:
            return self._default

        normalized = name.strip().lower().replace(" ", "")

        for domain in self._creators:
            enum_name = domain.name.lower()
            enum_value = domain.value.lower().replace(" ", "")

            if normalized in (enum_name, enum_value):
                return domain

        return self._default   
    
    def register(self,
        domain: DomainType,
        creator: Callable[[], BaseApplicationStrategy]) -> None:
        self._creators[domain] = creator

    def create(self, 
        domain: DomainType) -> BaseApplicationStrategy:
        return self._resolve_creator(domain)()

    def _resolve_creator(self,
        domain: DomainType) -> Callable[[], BaseApplicationStrategy]:
        return self._creators.get(domain, self._creators[self._default])

    def available(self) -> tuple[Enum, ...]:
        return tuple(self._creators.keys())
    
    def get_strategy(self,
        name: str | None) -> BaseApplicationStrategy:
        creator = self.get_creator(name)
        return creator()    

    def get_creator(self, name: str | None) -> Callable[[], BaseApplicationStrategy]:
        return self._resolve_creator(self._parse_domain(name))

File: angular-app/src/app/ddd-core/application/strategies/factories/strategy-application-factory.ts
import { ClaudeStrategy, GeminiStrategy, GroqStrategy, LangChainStrategy, OpenAIStrategy } from '..';
import { BaseDomain } from '../../../domain/abstracts/base-domain';
import { BaseApplicationStrategy, DomainConstructor } from '../abstracts/base-application-strategy';

export class StrategyApplicationFactory {
  private creators: Record<string, () => BaseApplicationStrategy> = {
    Groq: () => new GroqStrategy(),
    Gemini: () => new GeminiStrategy(),
    OpenAI: () => new OpenAIStrategy(),
    Claude: () => new ClaudeStrategy(),
    LangChain: () => new LangChainStrategy(),
  };

constructor(
    creators: Record<string, () => BaseApplicationStrategy> | null = null,
    public readonly defaultDomain: string = 'Groq'
  ) {
    if (creators) {
      Object.assign(this.creators, creators);
    }
    if (!this.creators[this.defaultDomain]) {
      throw new Error('Default strategy must exist in creators');
    }
  }

  private static _sanitize(value: string): string {
    return value.trim().toLowerCase().replace(/_/g, '').replace(/ /g, '');
  }

  public availableDomainKeys(): string[] {
    return Object.keys(this.creators);
  }

  public getDomainKey(value: string | DomainConstructor<BaseDomain> | null): string | null {
    const parsed = this.parseDomain(value);
    if (!parsed) return null;

    for (const [key, creator] of Object.entries(this.creators)) {
      const domainCtor = creator().domainClass as DomainConstructor<BaseDomain>;
      if (domainCtor === parsed) {
        return key;
      }
    }

    return null;
  }

  parseDomain(value: string | DomainConstructor<BaseDomain> | null): DomainConstructor<BaseDomain> | null {
    if (!value) return null;

    if (typeof value === 'function' && 'prototype' in value) {
      for (const creator of Object.values(this.creators)) {
        const domainCtor = creator().domainClass as DomainConstructor<BaseDomain>;
        if (domainCtor === value) {
          return domainCtor;
        }
      }
      return null;
    }

    const normalized = StrategyApplicationFactory._sanitize(value);
    for (const creator of Object.values(this.creators)) {
      const domainCtor = creator().domainClass;
      const aliases = [
        StrategyApplicationFactory._sanitize(domainCtor.name),
        StrategyApplicationFactory._sanitize(domainCtor.getDomainName()),
      ];
      if (aliases.includes(normalized)) {
        return domainCtor;
      }
    }
    return null;
  }

  register(domain: DomainConstructor<BaseDomain>, creator: () => BaseApplicationStrategy): void {
    this.creators[domain.name] = creator;
  }

  create(domain: DomainConstructor<BaseDomain>): BaseApplicationStrategy {
    return this._resolveCreator(domain)();
  }

  private _resolveCreator(domain: DomainConstructor<BaseDomain>): () => BaseApplicationStrategy {
    return this.creators[domain.name] || this.creators[this.defaultDomain];
  }

  available(): DomainConstructor<BaseDomain>[] {
    return Object.values(this.creators).map((creator) => creator().domainClass);
  }

  getStrategy(value: string | DomainConstructor<BaseDomain> | null): BaseApplicationStrategy {
    return this.getCreator(value)();
  }

  getCreator(value: string | DomainConstructor<BaseDomain> | null): () => BaseApplicationStrategy {
    const domain = this.parseDomain(value);
    if (!domain) {
      throw new Error(`Provider não suportado: ${String(value)}`);
    }
    return this._resolveCreator(domain);
  }
}
